framework/include/reporters/ReporterData.h

```diff
@@ -0,0 +1,319 @@
+   */
+  void finalize(const std::string & object_name);
+
+  /**
+   * At the end of a timestep this method is called to copy the values back in time in preparation
+   * for the next timestep.
+   *
+   * See FEProblemBase::advanceState
+   */
+  void copyValuesBack();
+
+  /**
+   * Return a set of undeclared names
+   *
+   * @see FEProblemBase::checkPostprocessors
+   */
+  std::set<ReporterName> getUndeclaredNames() const;
+
+  /**
+   * Writes all Reporter values to the supplied JSON node for output to a file.
+   *
+   * @see JSONOutput
+   */
+  void store(nlohmann::json & json) const;
+
+  /**
+   * Perform integrity check for get/declare calls
+   */
+  void check() const;
+
+  /**
+   * Return true if the supplied mode exists in the produced Repoter values
+   *
+   * @see JSONOutput.C/h
+   */
+  bool hasReporterWithMode(const ReporterMode & mode) const;
+
+private:
+  /// For accessing the restart/recover system, which is where Reporter values are stored
+  MooseApp & _app;
+
+  /**
+   * Helper method for creating the necessary RestartableData for Reporter values.
+   * @tparam T The desired C++ type for the Reporter value
+   * @param reporter_name Object/data name for the Reporter value
+   * @param declare Flag indicating if the ReporterValue is being declared or read. This flag
+   *                is passed to the existing MooseApp restart/recover system that errors if a
+   *                value is declared multiple times.
+   */
+  template <typename T>
+  ReporterState<T> & getReporterStateHelper(const ReporterName & reporter_name, bool declare);
+  friend class VectorPostprocessorInterface;
+
+  /**
+   * Helper method for returning the ReporterContextBase object, if it exists.
+   * @param reporter_name Object/data name for the Reporter value
+   */
+  const ReporterContextBase *
+  getReporterContextBaseHelper(const ReporterName & reporter_name) const;
+
+  /**
+   * Helper for registering data with the MooseApp to avoid cyclic includes
+   */
+  RestartableDataValue & getRestartableDataHelper(std::unique_ptr<RestartableDataValue> data_ptr,
+                                                  bool declare) const;
+
+  /// The ReporterContext objects are created when a value is declared. The context objects
+  /// include a reference to the associated ReporterState values. This container stores the
+  /// context object for each Reporter value.
+  ///
+  /// This container must be consistently sorted across processors to achieve a consistent iteration
+  /// order for parallel operations as well as for output. Objects are inserted based
+  /// on the combined ReporterName. A std::list was selected to achieve fast insert times
+  /// without re-allocating. A std::vector can also be used, but this was more fun. Do not use a
+  /// std::set, like I did initially, since the order is actually differs in each rank because it
+  /// is sorted by the pointer which causes all sorts of problems.
```

Comment:
🤣

Reply:
I believe here you could actually use a std::map with a custom sorting method on `ReporterContextBase`.

This could simplify all of your getters to use the standard map find instead of doing a linear search through a vector.